### PR TITLE
[Inference/Feat] Add kvcache quantization support for FlashDecoding

### PR DESCRIPTION
**Pull request by Courtesy-Xs**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by Courtesy-Xs** [comment]
image

File: extensions/csrc/funcs/unary_functor.h [hunk 1]
```diff
@@ -15,21 +15,6 @@
 namespace colossalAI {
 namespace funcs {
 
-template <typename T>
```
Comment (yuehuayingxueluo) [inline extensions/csrc/funcs/unary_functor.h:18]: Why do you want to delete this?

File: extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu [hunk 2]
```diff
@@ -174,13 +174,13 @@ void context_kv_cache_memcpy(
...
-            key,
-            value,
-            key_cache,
-            value_cache,
-            sequence_lengths,
-            cu_seqlens,
-            block_tables,
-            max_seq_len_in_batch
-        );)
+    key,
```
Comment (yuehuayingxueluo) [inline extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu:177]: There is an indentation issue here.

File: extensions/csrc/kernel/cuda/flash_decoding_attention_kernel.cu [hunk 3]
```diff
@@ -34,11 +33,25 @@ constexpr unsigned int nextHighestPowerOf2(unsigned int v) {
   return v;
 }
 
+template <typename T>
```
Comment (yuehuayingxueluo) [inline extensions/csrc/kernel/cuda/flash_decoding_attention_kernel.cu:36]: Why do you want to move 'zero' operation here?

**Review by yuehuayingxueluo — COMMENTED** [review]
I think you can add quantitative test cases to our flash attn unit tests.

**Review by SunflowerAries — APPROVED** [review]
LGTM

File: extensions/csrc/kernel/cuda/flash_decoding_attention_kernel.cu (hunk 3 above)
Comment (Courtesy-Xs) [reply extensions/csrc/kernel/cuda/flash_decoding_attention_kernel.cu:36]: Just be used here

File: extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu (hunk 2 above)
Comment (Courtesy-Xs) [reply extensions/csrc/kernel/cuda/context_kv_cache_memcpy_kernel.cu:177]: Next PR I will fix it

**Review by yuehuayingxueluo — APPROVED** [review]
LGTM

**Review by yuanheng-zhao — APPROVED** [review]
LGTM

**Pull request merged**